### PR TITLE
Generalize training

### PR DESCRIPTION
**Pull request by Ingvarstep**
Overall, it was generalised training by the following changes:

- Switched to focal loss instead of binary cross-entropy loss and constant weighting. With specific parameters of focal loss, it will become binary cross-entropy; in this case, alpha and gamma should be equal to 0;
- Added more scheduler types, like linear, polynomial, constant;
- Added support of different weight decays for the encoder and other components of a model;

### CODE REVIEW

**Review by urchade — COMMENTED** [review]
Looks good
